### PR TITLE
Webhook signature docs

### PR DESCRIPTION
**Pull request by rishavbhowmik**
- Updated screen short with, 'webhook secret' changes
- `## Verify webhook signature`
- `## Preventing replay attacks`
- New sample codes

### CODE REVIEW

File: api-reference/api-introduction/webhooks.md [hunk 1]
```diff
@@ -39,34 +41,352 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+```
+
+{% endtab %}
+{% endtabs %}
+
+### Verify signature manually
+
+Imagekit webhook signature follows the following format:
+
+```js
```
Comment (imagekitio) [inline api-reference/api-introduction/webhooks.md:87]: We need to mention that it comes in header. Be verbose. Show it in context.

File: api-reference/api-introduction/webhooks.md [hunk 2]
```diff
@@ -39,34 +41,352 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+timestamp = 't=1655795539264'.split('=')[1]
+// '1655795539264'
+```
+
+Step 3: Extract HMAC hash from the signature.
+
+```js
+hmac = 'v1=b6bc2aa82491c32f1cbef0eb52b7ffaa51467ea65a03b5d4ccdcfb9e0941c946'.split('=')[1]
```
Comment (imagekitio) [inline api-reference/api-introduction/webhooks.md]: dont' use work hmac here. say it `signature`

File: api-reference/api-introduction/webhooks.md [hunk 3]
```diff
@@ -39,34 +41,352 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+
+## Preventing replay attacks
+
+When an attacker intercepts a webhook request, he can replay the request multiple times with a valid signature.
+
+To mitigate this, Imagekit webhook signature contains a timestamp. The timestamp is generated right before the webhook request is sent to the server.
+
+For retriving this timestamp, the `verify` method in ImagekitSDK returns timestamp & parsed event object.
+
+If timestamp is within the tolerance limit, the request can be considered as valid, else the request must be stalled.
```
Comment (imagekitio) [inline api-reference/api-introduction/webhooks.md]: Must be stalled? or ignored?

File: api-reference/api-introduction/webhooks.md [hunk 4]
```diff
@@ -39,34 +41,352 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+
+> Optionally, a stronger approach is to use a nonce to prevent replay attacks. You can use Event ID as nonce, it is guaranteed to be unique across all events.
```
Comment (imagekitio) [inline api-reference/api-introduction/webhooks.md]: Write it as a normal paragraph.

File: api-reference/api-introduction/webhooks.md [hunk 5]
```diff
@@ -39,34 +41,352 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+const WEBHOOK_ENDPOINT = "/webhook";
+const WEBHOOK_SECRET = "whsec_EvCLlGyctj2WyIMW2r62m3JnyN4SXMsY"; // Copy from Imagekit dashboard
+const WEBHOOK_EXPIRY_DURATION = 60 * 1000; // 60 seconds
+
+// Server configs
+const PORT = 8081;
+
+const app = express();
+
+app.post("/webhook", express.raw({ type: "*/*" }), (req, res) => {
```
Comment (imagekitio) [inline api-reference/api-introduction/webhooks.md]: `{ type: "*/*" }` or `application/json`. We saw that that req.body was still buffer with `application/json` right?

File: api-reference/api-introduction/webhooks.md [hunk 6]
```diff
@@ -39,34 +41,352 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+  }
+
+  // Handle webhook
   switch (event.type) {
-    case 'video.transformation.accepted':
-      // It is triggered when a new video transformation request is accepted for processing. You can use this for debugging purposes.
+    case "video.transformation.accepted":
+      console.log("Video transformation request accepted");
       break;
-    case 'video.transformation.ready':
```
Comment (imagekitio) [inline api-reference/api-introduction/webhooks.md:55]: Please revert this change. They are more verbose and in line with our docs.

File: api-reference/api-introduction/webhooks.md [hunk 7]
```diff
@@ -39,34 +41,352 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+
+      // Check if webhook has expired
+      if (timestamp.getTime() + WEBHOOK_EXPIRY_DURATION < Date.now()) {
+        // Stall webhook
+        return res.status(401).send("Webhook signature expired");
+      }
+      
+      // Handle webhook
+      switch (event.type) {
+        case "video.transformation.accepted":
```
Comment (imagekitio) [inline api-reference/api-introduction/webhooks.md]: Please keep comments as Express.

**Review by imagekitio — CHANGES_REQUESTED** [review]

File: api-reference/api-introduction/webhooks.md [hunk 8]
```diff
@@ -39,34 +41,352 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+Step 4: Compute the HMAC hash using the webhook secret, signature timestamp and raw request body.
+
+- HAMC key is the webhook secret.
+- HMAC payload is formed by concatenating the timestamp(as numeric string), character `.` and raw request body (encoded as UTF8 string).
+- Use SHA256 algorithm to compute the HMAC hash.
+- HMAC hash is hex encoded.
+
+```js
+computeHmac = Hmac('sha256')
+  .Key(webhookSecret)
```
Comment (imagekitio) [inline api-reference/api-introduction/webhooks.md]: What is `Hmac` here? It should be simple
```js
const createHmac = require("crypto").createHmac;
createHmac("sha256", webhookSecret).update(timestamp + '.' + rawRequestBody).digest("hex");
```

**Review by imagekitio — CHANGES_REQUESTED** [review]

File: api-reference/api-introduction/webhooks.md [hunk 9]
```diff
@@ -39,21 +41,172 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+  try {
+    webhookResult = imagekit.verifyWebhookEvent(rawBody, signature, WEBHOOK_SECRET);
+  } catch (e) {
+    // Failed to verify webhook
+    return res.status(401).send(`Webhook error: ${e.message}`);
+  }
+  const { timestamp, event } = webhookResult;
+
+  // Check if webhook has expired
+  if (timestamp.getTime() + WEBHOOK_EXPIRY_DURATION < Date.now()) {
```
Comment (rishavbhowmik) [inline api-reference/api-introduction/webhooks.md]: timestamp is now a number

File: api-reference/api-introduction/webhooks.md [hunk 10]
```diff
@@ -62,11 +215,200 @@ app.post('/webhook', express.json({type: 'application/json'}), (request, respons
...
+      try {
+        webhookResult = imagekit.verifyWebhookEvent(rawBody, signature, WEBHOOK_SECRET);
+      } catch (e) {
+        // Webhook verification failed
+        return res.status(401).send(`Webhook error: ${e.message}`);
+      }
+      const { timestamp, event } = webhookResult;
+
+      // Check if webhook has expired
+      if (timestamp.getTime() + WEBHOOK_EXPIRY_DURATION < Date.now()) {
```
Comment (rishavbhowmik) [inline api-reference/api-introduction/webhooks.md]: timestamp is number

File: api-reference/api-introduction/webhooks.md [hunk 11]
```diff
@@ -62,11 +215,200 @@ app.post('/webhook', express.json({type: 'application/json'}), (request, respons
...
+    } catch (e) {
+      // Webhook signature verification failed
+      res.writeHead(401, `Webhook error: ${e.message}`);
+      res.end();
+      return;
+    }
+    const { timestamp, event } = webhookResult;
+
+    // Check if webhook has expired
+    if (timestamp.getTime() + WEBHOOK_EXPIRY_DURATION < Date.now()) {
```
Comment (rishavbhowmik) [inline api-reference/api-introduction/webhooks.md]: timestamp is number

**Review by rishavbhowmik — COMMENTED** [review]
timestamp is number not Date object

File: api-reference/api-introduction/webhooks.md [hunk 12]
```diff
@@ -39,21 +41,172 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
 | createdAt | `string` | Timestamp of the event in ISO8601 format.|
 | data      | `JSON`   | Actual payload of event in JSON format.  |
 
-### Sample code
+## Verify webhook signature
+
+Webhook endpoints are publicly accessible, therefore is important filter out malicious requests. We recommend that you use webhook signature to verify the authenticity of the webhook request.
```
Comment (rishavbhowmik) [inline api-reference/api-introduction/webhooks.md]: Grammer! 'important to filter'

File: api-reference/api-introduction/webhooks.md [hunk 13]
```diff
@@ -39,21 +41,172 @@ All webhook bodies are JSON encoded. The schema of the body may differ based on
...
+
+To achieve this, every imagekit webhook endpoint comes with a signature in the request header `x-ik-signature`. To generate this signature, webhook secret is required, you can find webhook secret in [Imagekit dashboard](https://imagekit.io/dashboard/developer/webhooks).
```
Comment (rishavbhowmik) [inline api-reference/api-introduction/webhooks.md]: Correction! 'every imagekit webhook request'

**Review by rishavbhowmik — COMMENTED** [review]
Fixes for `## Verify webhook signature`

**Pull request merged**